### PR TITLE
Implement `on_killed` support for tasks during DAG run timeout

### LINKED ISSUE #41036: Tasks that are running when a DAG run reaches its timeout should optionally support `on_killed`
### Description

The current functionality for task instances that are in the running state when a DAG run reaches its timeout is controversial (see #30264). It'd be great if users could optionally allow tasks' `on_kill` method to be called at this time.

### Use case/motivation

Some users would like for externally running workloads (e.g. snowflake, emr, bigquery, etc etc) to stop executing when a DAG run times out.

### Related issues

#30264

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by MRLab12**
---

This PR addresses issue #41036 by adding support for the `on_killed` callback on tasks that are still running when a DAG run reaches its timeout.

Key changes:

1. Added a new configuration option `call_on_kill_on_dagrun_timeout` to control whether tasks should be killed when a DAG run times out (default: False).
2. Updated logic in `_schedule_dag_run` to call task `on_kill` if  `call_on_kill_on_dagrun_timeout` is enabled.

closes: https://github.com/apache/airflow/issues/41036

Testing:
- Added unit test for the new timeout handling logic
- Verified that the new configuration option works as expected

### CODE REVIEW

**Comment by MRLab12** [comment]
I'm having some issues with the test I added. Looking at other tests in `test_scheduler_job` I see that `session.refresh(dr)` is used before asserting the dag run state. In my test it causes a missing state error. Has anyone encountered this before?

Also. I want to check that the `on_kill` method of `BashOperator` is called, but haven't gotten the Mock to work and it is currently failing. Looking into it.

I wanted to get this draft PR up to check that I did the correct changes. Hopefully I understood the issue correctly.

**Comment by MRLab12** [comment]
Apologies for the merge mess!

File: airflow/api_connexion/schemas/dag_schema.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ class DAGDetailSchema(DAGSchema):
     template_searchpath = fields.String(dump_only=True)
     render_template_as_native_obj = fields.Boolean(dump_only=True)
     last_loaded = fields.DateTime(dump_only=True, data_key="last_parsed")
+    call_on_kill_on_dagrun_timeout = fields.Boolean(dump_only=True)
```
Comment (eladkal) [inline airflow/api_connexion/schemas/dag_schema.py:116]: What would be the scenario where dagrun/task timeout and you wouldn't want to execute the `on_kill`?

Comment (MRLab12) [reply airflow/api_connexion/schemas/dag_schema.py:116]: From what I understood of the issue #41036 , OP is requesting a flag for allowing tasks to run `on_kill` when a dagrun reaches timeout. The use case given in the issue is that "Some users would like for externally running workloads (e.g. snowflake, emr, bigquery, etc etc) to stop executing when a DAG run times out." . 

This feature would add the flexibility to control this behavior as the user wants. I might have misunderstood the intention though, happy to make changes if the changes i made weren't what this issue needs.

Maybe @RNHTTR can help clarify.

**Pull request closed without merging**